### PR TITLE
Add possibility to set filename.

### PR DESCRIPTION
**Pull request by Primajin**
#### Setting the names for generated files

By default the file names follow a certain pattern. You can override this by setting the `fileName` property in your `backstop.json` to something other than `null`.
Use these tokens and remember to create unique files and set the file extension to `.png`

| Token | Replacement |
| ----- | ----------- |
| `__scenario_index__` | **0**_0_header_0_phone.png |
| `__counter__` | 0_**0**_header_0_phone.png |
| `__cleanedSelectorName__` | 0_0_**header**_0_phone.png |
| `__viewport_index__` | 0_0_header_**0**_phone.png |
| `__vp.name__` | 0_0_header_0_**phone**.png |

default sequence is:

```js
scenario_index + '_' + i + '_' + cleanedSelectorName + '_' + viewport_index + '_' + vp.name + '.png'
```

### CODE REVIEW

**Comment by Primajin** [comment]
#137 Might fix #101 :wink:

**Comment by Primajin** [comment]
We are using BackstopJS in Jenkins CI for regression testing, everytime someone fiddles around with the styles to ensure nothing breaks somewhere else you're not thinking about.
Furthermore we use it for automated screenshotting to get a feeling for the "evolution" of our pages over the feature iterations.
Having filenames that correspond to the nice named scenarios rather than sometimes very weird (or duplicated) css selectors helps us to identify the screenshots more easily and clearly.
The possibility to have clearer names will greatly decrease effort as we also have people looking at the files, that are technically less into selectors and coding :wink: 
Also it could be very nice to be able to just swap tokens or use hyphenation instead of underscore.
I ran various tests with it and reports were fine as long as filenames stayed unique.
Sure you can mess it up, but thats the same thing as messing up selectors or the paths.

**Comment by garris** [comment]
Ok. I've thought about this a bit.  I do think it's legitimate to allow the user to reformat the filename.   But I would suggest implementing it a little differently inside the app.  Here are my thoughts...

- do it globally in the config, add a property `fileNameTemplate`
- do it as an ES6 or a mustache template e.g. `{scenarioIndex}_{scenarioLabel}_{selectorIndex}_{selectorLabel}_{viewportIndex}_{viewportLabel}`
- add this as a default property to `gulp/paths.js` and allow the initial value to be overridden.
- the default property should retain the current format because changing it now would break existing tests for existing users.
- I'd do all the chaining and clean for all file systems last e.g. 
```
var fileName = paths.fileNameTemplate.replace(/{key}/, value)
  .replace(/{key}/, value)
  .replace(/{key}/, value)
  .replace(/{key}/, value)
  .replace(/{key}/, value)
  .replace(/{key}/, value)
  .replace(/[^a-z0-9_\-]/gi,'')
```

**Comment by garris** [comment]
I'd also check for a `.png` suffix and just add it for the user if they leave it off.

**Comment by Primajin** [comment]
Allright cool!

**Comment by klausbayrhammer** [comment]
Awesome idea! Really looking forward to this! Just let me know if there is anything I can help you with!

**Comment by garris** [comment]
@Primajin Thanks for getting this topic going.  I also wanted to check in and see if you were making any progress or wanted some assistance.  We all have day jobs so -- no pressure!  Just please let us know when you get a chance.  Thanks!

**Comment by garris** [comment]
Additional note: we will likely also need a representation of the config file path to ensure different configs that point to the same resource directories do not clobber existing same-name files.

**Comment by Primajin** [comment]
Sorry guys, I was not around or some time. I will have a look into it on the weekend.
@garris could you rephrase / specify your last post, I didn't quite understood it.

**Comment by garris** [comment]
Hi @Primajin, thanks for following up.   

I realized there is a currently a potential for file corruption/clobbering if a user creates multiple config files that share the same resource directories.

Here's an example:

Imagine two config files named `config1.json` and `config2.json`.
- `config1.json` has a single scenario labeled "FEED" and a single viewport named "PHONE". It points to a local instance of a user's feed.
- `config2.json` is identical except the scenario's URL points to a staging instance.

If your file naming template string was `{scenarioIndex}_{scenarioLabel}_{selectorIndex}_{selectorLabel}_{viewportIndex}_{viewportLabel}` and you were to do a `gulp reference --backstopConfigFilePath=./config1.json ` you'd create a reference screen capture named `0_FEED_0_body_0_PHONE.png`.  At this point running `gulp test --backstopConfigFilePath=./config1.json` works fine.

The issue is created when running `gulp reference --backstopConfigFilePath=./config2.json `.  Doing that would clobber  `0_FEED_0_body_0_PHONE.png` with the staging bitmap.  Running `gulp test --backstopConfigFilePath=./config1.json` after that point would give false results.

I think this is a significant enough issue that it needs to be addressed in the default configuration.

To prevent the issue you would need to add a config identifier... `{configFile}_{scenarioIndex}_{scenarioLabel}_{selectorIndex}_{selectorLabel}_{viewportIndex}_{viewportLabel}`

The `configFile` could probably be hashed (and/or shortened and/or explicitly labeled) so that files are still human browsable.

Hope this helps!  Please let me know if you have any better solutions or other ideas.  Thanks!

**Comment by Primajin** [comment]
Allright I'm on it. Currently there is still a bug in there that I try to squash - I keep you posted

**Comment by Primajin** [comment]
BTW: Do you prefer to have just a single commit in the PR when all is done?

**Comment by Primajin** [comment]
OK I might need some help here, because my JavaScript Knowledge is limited.
So I though I broke something with my changes in my branch / PR - but when I check out your master I can't generate files either..

CasperJS: TypeError: 'undefined' is not a function (evaluating 'this.evaluate(function() {return document.querySelector(o).outerHTML;})')

Can you verify this?

**Comment by garris** [comment]
@Primajin Many apologies.  There were a few commits to master which should have gone to a different branch.  These have been reverted.  Master should now be working.   Thank you.

**Comment by Primajin** [comment]
Allright rebased my fork.

`CasperJS: Screenshots for tablet_h (1024x768)
CasperJS: Using filename pattern: Default_0_MyHomepage_0_header_2_tablet_h.png
CasperJS: Using filename pattern: Default_0_MyHomepage_1_main_2_tablet_h.png
CasperJS: Using filename pattern: Default_0_MyHomepage_2_bodybs-docs-featurettenth-of-type1_2_tablet_h.png
CasperJS: Using filename pattern: Default_0_MyHomepage_3_bodybs-docs-featurettenth-of-type2_2_tablet_h.png
CasperJS: Using filename pattern: Default_0_MyHomepage_4_footer_2_tablet_h.png
CasperJS: Using filename pattern: Default_0_MyHomepage_5_body_2_tablet_h.png
CasperJS: Comparison config file updated.`

**Comment by garris** [comment]
Thank you @primajin.  I am on holiday.  Will review later this week and follow up. Cheers.

**Comment by garris** [comment]
ok @Primajin, good work!  

Heres comments -- please forgive me if there are errors or miscommunications in this review.  Just ask any questions if I'm being unclear or confusing.  I wanted to get back to you as soon as possible...

----
on this line...

https://github.com/Primajin/BackstopJS/commit/f6de346ec5d6b3b1cf9f4e358a231c7865af4590#diff-8854ff0ca17d960be935d14ef5decc3fR31

`genConfigPath` is *almost* right.  `genConfigPath` is an internal pointer to a _cached_ version of the current config file.  We want the path the actual currently selected config used to run the tests.

The bad news is that path is not currently available to `genConfig` — the good news is that it is easy to add.  Here is where we make the cached copy...
     https://github.com/Primajin/BackstopJS/blob/master/gulp/tasks/_init_.js#L6
Above that line do this...
    `config.configFilePath = paths.activeCaptureConfigPath || '';`

Then here...
https://github.com/Primajin/BackstopJS/blob/master/capture/genBitmaps.js#L31
you can...
`var config_name = config.name || config.configFilePath.hashCode();`

I think that gets us going.   Here are more notes...

Is it necessary to echo out every file we write?  
https://github.com/Primajin/BackstopJS/blob/master/capture/genBitmaps.js#L190
I don't think users need to know this.  FWIW: I also have a task to clean the other unnecessary stuff being spit out on the CLI (so, don't feel bad)

Might want to tweak `.hashCode()` so that it never generates a leading ‘minus’ sign character — the dash could complicate things considering that it will be used for naming files and will be seen in reports.

Before the next release I am planning on pruning/simplifying the config boilerplate config file and I am inclined to remove `fileNameTemplate` from the boilerplate -- I only want stuff in there that _most_ users will need.   In any case, we are going to want some decent documentation around this.

We also need clear upgrade documentation around this feature and possibly a compatibility utility — e.g. a gulp task which will take a file (or directory) path and add a `fileNameTemplate` node to any JSON file containing a `scenarios` node (to ensure we are modifying a backstopjs file).  The added node should explicitly set the file naming to `{scenarioIndex}_{selectorIndex}_{selectorLabel}_{viewportIndex}_{viewportLabel}`  so that newly created test samples are compatible with the older format.

There you have it. Great job so far!

**Comment by Primajin** [comment]
Sorry guys I will continue on this asap, had so much stuff to do in the last weeks ;)

**Comment by garris** [comment]
thanks @Primajin!   No worries.

**Comment by Primajin** [comment]
I will try to work more on this in the next days, but I've seen that @Pofigizm already sorted some things out in #197 ... maybe we can join forces?

**Comment by Pofigizm** [comment]
Feature "set filename" is very useful for me and my project. I'm ready to help.
In my PR I only revert default settings, as @garris asks.

**Comment by Primajin** [comment]
Allright cool, then I will pull in your changes.

**Comment by Primajin** [comment]
@Pofigizm I added your changes. I'm currenlty running into an issue that CLI exits with 0 (green) although there are errors and `"cliExitOnFail": true` is set in `backstop.json`.

**Comment by Pofigizm** [comment]
I haven't problem in [my fork](https://github.com/Pofigizm/BackstopJS/tree/filename) which i use everyday. 
Maybe because https://github.com/garris/BackstopJS/commit/b78be9ba707d2820abc86bf6b1767c4587d52f77 doesn't exist there?

**Comment by Primajin** [comment]
So @garris Is there anything else we can do to get this PR done?

**Comment by garris** [comment]
@primajin I am having some trouble following this thread.  Could you recap your status. 

- what has been completed
- what is left to do 
- are there any issues blocking you?

**Comment by Primajin** [comment]
@garris So I tested everything thoroughly but I'm still facing the problem that the build exits with 0 although there were mismatches. I need to check if this is also on the master or other branches to see where this comes from. Everything else should work as expected.

So this can't be merged into master right now, as it would not "crash as expected" ;)

**Comment by Primajin** [comment]
Finally! I rebased my fork to the latest upstream version and now it is finally "failing as desired" (so no more false positive builds).

I also added a hint to the error message, so that the CLI user can see the same as the browser compare user: How far the resemblance was according to the set threshold:

`ERROR (1.81% / 0.06%): Molecules - Header (Page) Molecules-HeaderPage_0_test-standardheader_2_tablet_h.png`

So the user knows that the threshold was 0.06% but the test rendering was 1.81% different to the reference. If the complete dimensions fail, there is also a hint for that.

Hope now everything is in order @garris , looking forward to your comments and or pull ;)
Best!

**Comment by garris** [comment]
That's great. I will try to take a look this weekend.  I am getting a little backed up on my BackstopJS todo's.

**Comment by Primajin** [comment]
So @garris Could you look at this?

**Comment by garris** [comment]
So sorry,  I've been renovating my house and had very little time this month.  I will review this soon.  Thanks for your patience.  

Regards.

**Comment by Primajin** [comment]
No worries, and be careful with the fuses :zap: ;)

**Comment by garris** [comment]
Hi @primajin,  I need to take another review pass -- but in general, this is looking good!  

I just created branch `set_filename`.   Would you please squash your commits and PR to the new branch when you get a chance?

**Comment by Primajin** [comment]
Allright I opened #247

**Comment by garris** [comment]
This was manually moved into version 2 and released. Thank you for the contribution!

**Comment by Primajin** [comment]
Wohoo 🎉
